File: editors/vscode/src/command-diagnostics.js
import * as path from "node:path";
import { readFileSync } from "node:fs";
import { integerInRange, protocol } from "./command-protocol.js";
import diagnosticMessages from "./diagnostics.generated.js";
import diagnosticContracts from "./diagnostic-contract.generated.js";

export function applyDiagnostics(api, collection, records, projectRoot, knownUris) {
  applyProjectedDiagnostics(collection, projectDiagnostics(api, records, projectRoot), knownUris);
}

/**
 * Project every record before replacing a collection. A source read/range
 * failure therefore leaves the previous snapshot intact instead of exposing
 * a partially projected or empty result.
 */
export function replaceDiagnostics(api, collection, records, projectRoot, knownUris) {
  const values = projectDiagnostics(api, records, projectRoot);
  const entries = new Map([...knownUris.values()].map((uri) => [uri.toString(), [uri, []]]));
  for (const entry of values.values()) entries.set(entry.uri.toString(), [entry.uri, entry.values]);
  collection.set([...entries.values()]);
  knownUris.clear();
  for (const entry of values.values()) knownUris.set(entry.uri.toString(), entry.uri);
}

function projectDiagnostics(api, records, projectRoot) {
  if (!Array.isArray(records)) throw protocol("invalid_diagnostics");
  const values = new Map();
  for (const record of records) {
    const diagnostic = toDiagnostic(api, record, projectRoot);
    if (!diagnostic) continue;
    const key = diagnostic.uri.toString();
    const entry = values.get(key) ?? { uri: diagnostic.uri, values: [] };
    entry.values.push(diagnostic.value);
    values.set(key, entry);
  }
  return values;
}

function applyProjectedDiagnostics(collection, values, knownUris) {
  collection.set([...values.values()].map((entry) => [entry.uri, entry.values]));
  for (const entry of values.values()) knownUris.set(entry.uri.toString(), entry.uri);
}

export function clearDiagnostics(collection, knownUris) {
  collection.clear?.();
  knownUris.clear();
}

function toDiagnostic(api, record, projectRoot) {
  if (!validDiagnosticRecord(record)) {
    throw protocol("invalid_diagnostic");
  }
  const fullPath = diagnosticPath(projectRoot, record.span.file);
  const open = api.workspace?.textDocuments?.find((document) =>
    document.uri?.scheme === "file" && path.normalize(document.uri.fsPath) === fullPath);
  if (open?.isDirty) return undefined;
  const text = sourceText(api, fullPath);
  const range = rangeForSpan(api, text, record.span);
  const uri = api.Uri.file(fullPath);
  const severityName = record.severity;
  const severity = api.DiagnosticSeverity?.[
    ({ error: "Error", warning: "Warning", information: "Information", hint: "Hint" })[severityName] ?? "Error"
  ] ?? api.DiagnosticSeverity?.Error;
  const value = new api.Diagnostic(
    range,
    diagnosticMessage(record.presentation, record.compatibility_message ?? record.code),
    severity
  );
  value.code = record.code;
  value.source = "recite";
  return { uri, value };
}

function sourceText(api, fullPath) {
  // Structured CLI spans describe the saved on-disk source snapshot. A clean
  // open document is equivalent to that snapshot and avoids a disk race;
  // dirty overlays are deliberately ignored so CLI diagnostics never attach
  // to unsaved text.
  const open = api.workspace?.textDocuments?.find((document) =>
    document.uri?.scheme === "file" && path.normalize(document.uri.fsPath) === fullPath);
  if (open && !open.isDirty && typeof open.getText === "function") return open.getText();
  try {
    return readFileSync(fullPath, "utf8");
  } catch (error) {
    throw protocol("diagnostic_source_unavailable", error.message);
  }
}

function rangeForSpan(api, text, span) {
  const start = positionForScalar(api, text, span.start, "invalid_diagnostic_start");
  if (!span.end) return new api.Range(start, start);
  const endPosition = advanceInclusive(text, span.end);
  const end = positionForScalar(api, text, endPosition, "invalid_diagnostic_end");
  if (end.line < start.line || end.line === start.line && end.character < start.character) {
    throw protocol("invalid_diagnostic_range");
  }
  return new api.Range(start, end);
}

function positionForScalar(api, text, position, errorCode) {
  const lines = text.split("\n");
  const lineIndex = position.line - 1;
  const line = lines[lineIndex]?.endsWith("\r") ? lines[lineIndex].slice(0, -1) : lines[lineIndex];
  if (line === undefined) throw protocol(errorCode);
  const scalarOffset = position.column - 1;
  const values = [...line];
  if (scalarOffset < 0 || scalarOffset > values.length) throw protocol(errorCode);
  const character = values.slice(0, scalarOffset).reduce((total, value) => total + value.length, 0);
  return new api.Position(lineIndex, character);
}

function advanceInclusive(text, position) {
  const lines = text.split("\n");
  const raw = lines[position.line - 1];
  const line = raw?.endsWith("\r") ? raw.slice(0, -1) : raw;
  if (line === undefined) throw protocol("invalid_diagnostic_end");
  const scalarOffset = position.column - 1;
  const values = [...line];
  if (scalarOffset < 0 || scalarOffset >= values.length) {
    throw protocol("invalid_diagnostic_end");
  }
  return { line: position.line, column: position.column + 1 };
}

export function validDiagnosticRecord(record) {
  return Boolean(record && keys(record, ["version", "code", "severity", "span", "presentation", "related", "help", "explanation", "compatibility_message"]) &&
    record.version === 1 && typeof record.code === "string" && /^[A-Z][A-Z0-9]*_[A-Z0-9]+$/u.test(record.code) &&
    ["error", "warning", "information", "hint"].includes(record.severity) && validSpan(record.span) &&
    validPresentation(record.presentation) && Array.isArray(record.related) &&
    record.related.every(validRelatedPresentation) &&
    (record.help === null || record.help === undefined || validPresentation(record.help)) &&
    (record.explanation === null || record.explanation === undefined || validExplanation(record.explanation)) &&
    (record.compatibility_message === null || record.compatibility_message === undefined ||
      typeof record.compatibility_message === "string"));
}

function validSpan(span) {
  return Boolean(span && keys(span, ["file", "start", "end"]) &&
    typeof span.file === "string" && span.file.length > 0 &&
    positive(span.start?.line) && positive(span.start?.column) &&
    (span.end === null || positive(span.end?.line) && positive(span.end?.column) &&
      (span.end.line > span.start.line || span.end.line === span.start.line &&
        span.end.column >= span.start.column)));
}

function validPresentation(presentation) {
  if (!presentation || !keys(presentation, ["id", "arguments"]) ||
      typeof presentation.id !== "string" ||
      !/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/u.test(presentation.id) ||
      !presentation.arguments || typeof presentation.arguments !== "object" ||
      Array.isArray(presentation.arguments) ||
      !Object.keys(presentation.arguments).every((name) => /^[a-z][a-z0-9_]*$/u.test(name)) ||
      !Object.values(presentation.arguments).every(validArgument)) return false;
  const definition = diagnosticContracts[presentation.id];
  if (!definition) return true;
  const expected = new Map(definition.arguments.map(({ name, type }) => [name, type]));
  return Object.keys(presentation.arguments).length === expected.size &&
    Object.entries(presentation.arguments).every(([name, argument]) =>
      expected.has(name) && validArgument(argument, expected.get(name)));
}

function validRelatedPresentation(related) {
  return Boolean(related && keys(related, ["span", "presentation"]) &&
    validSpan(related.span) && validPresentation(related.presentation));
}

function validExplanation(explanation) {
  return Boolean(explanation && keys(explanation, ["meaning", "common_causes", "remediation"]) &&
    validPresentation(explanation.meaning) &&
    Array.isArray(explanation.common_causes) && explanation.common_causes.every(validPresentation) &&
    Array.isArray(explanation.remediation) && explanation.remediation.every(validPresentation));
}

function validArgument(argument, expectedType) {
  return Boolean(argument && typeof argument === "object" && keys(argument, ["type", "value"]) &&
    ["string", "integer", "float", "boolean"].includes(argument.type) &&
    (!expectedType || argument.type === expectedType) &&
    (argument.type === "string" ? typeof argument.value === "string" :
      argument.type === "integer" ? integerInRange(argument.value, "-9223372036854775808", "9223372036854775807") :
        argument.type === "float" ? typeof argument.value === "number" && Number.isFinite(argument.value) :
          typeof argument.value === "boolean"));
}

function diagnosticMessage(presentation, fallback) {
  const definition = diagnosticMessages[presentation.id];
  if (!definition) return fallback;
  const template = definition.template;
  const expected = new Map(definition.arguments.map(({ name, type }) => [name, type]));
  if (Object.keys(presentation.arguments).length !== expected.size ||
      Object.entries(presentation.arguments).some(([name, argument]) =>
        !expected.has(name) || !validArgument(argument, expected.get(name)))) return fallback;
  let unresolved = false;
  const rendered = template.replace(/\{\$([a-zA-Z][a-zA-Z0-9_-]*)\}/gu, (_, name) => {
    const argument = presentation.arguments[name];
    if (!argument) unresolved = true;
    return argument ? String(argument.value) : `{${name}}`;
  });
  return unresolved ? fallback : rendered;
}

function diagnosticPath(root, file) {
  const fullPath = path.isAbsolute(file) ? path.normalize(file) : path.resolve(root, file);
  const relative = path.relative(root, fullPath);
  if (!relative || relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw protocol("diagnostic_path_outside_project");
  }
  return fullPath;
}

function positive(value) {
  return Number.isSafeInteger(value) && value > 0;
}

function keys(value, expected) {
  return JSON.stringify(Object.keys(value).sort()) === JSON.stringify(expected.slice().sort());
}

File: editors/recite-neovim/lua/recite/command_diagnostics.lua
local protocol = require("recite.command_protocol")
local diagnostic = require("recite.diagnostic_protocol")

local M = {}

local function absolute(root, file)
  if file:sub(1, 1) == "/" then return vim.fn.fnamemodify(file, ":p") end
  return vim.fn.fnamemodify(root .. "/" .. file, ":p")
end

local function read_source(path)
  local handle, error = io.open(path, "rb")
  if not handle then return nil, error end
  local value = handle:read("*a")
  handle:close()
  return value
end

local function line_for(text, line)
  local start = 1
  local current = 1
  while current < line do
    local newline = text:find("\n", start, true)
    if not newline then return nil end
    start = newline + 1
    current = current + 1
  end
  local finish = text:find("\n", start, true)
  local value = text:sub(start, finish and finish - 1 or #text)
  if value:sub(-1) == "\r" then value = value:sub(1, -2) end
  return value
end

local function scalar_to_byte(line, column)
  local raw = protocol.integer_raw(column)
  if not raw or not protocol.integer_in_range(raw, "1", "18446744073709551615") then return nil end
  local target = tonumber(raw) - 1
  if not target or target < 0 then return nil end
  if target == 0 then return 0 end
  local index = 1
  local scalar = 0
  while index <= #line and scalar < target do
    local byte = line:byte(index)
    local width = 1
    if byte >= 0xF0 then width = 4
    elseif byte >= 0xE0 then width = 3
    elseif byte >= 0xC0 then width = 2 end
    if index + width - 1 > #line then return nil end
    index = index + width
    scalar = scalar + 1
  end
  if scalar ~= target then return nil end
  return index - 1
end

local function buffer_for(path)
  local bufnr = vim.fn.bufnr(path)
  if bufnr == -1 then
    bufnr = vim.fn.bufadd(path)
  end
  if not bufnr or bufnr == -1 then return nil end
  return bufnr
end

local function dirty(bufnr)
  return bufnr and vim.api.nvim_buf_is_valid(bufnr) and vim.bo[bufnr].modified
end

local function diagnostic_message(record)
  local rendered = diagnostic.render(record.presentation)
  return rendered or record.compatibility_message or record.code
end

local function to_diagnostic(record, project_root)
  if not diagnostic.valid(record) then error(protocol.error("invalid_diagnostic")) end
  local path = absolute(project_root, record.span.file)
  local root = vim.fn.fnamemodify(project_root, ":p"):gsub("/+$", "")
  if root == "" then root = "/" end
  local inside = root == "/" and path:sub(1, 1) == "/"
    or path == root or path:sub(1, #root + 1) == root .. "/"
  if not inside then
    error(protocol.error("diagnostic_path_outside_project"))
  end
  local source, read_error = read_source(path)
  if not source then error(protocol.error("diagnostic_source_unavailable", read_error)) end
  local start_line = protocol.integer_raw(record.span.start.line)
  local start_line_number = tonumber(start_line)
  local line = line_for(source, start_line_number)
  if not line then error(protocol.error("invalid_diagnostic_start")) end
  local start_col = scalar_to_byte(line, record.span.start.column)
  if not start_col then error(protocol.error("invalid_diagnostic_start")) end
  local end_col = start_col
  local end_line_number = start_line_number
  if record.span["end"] ~= nil and record.span["end"] ~= vim.NIL then
    end_line_number = tonumber(protocol.integer_raw(record.span["end"].line))
    local end_line = line_for(source, end_line_number)
    if not end_line then error(protocol.error("invalid_diagnostic_end")) end
    local end_start = scalar_to_byte(end_line, record.span["end"].column)
    if not end_start then error(protocol.error("invalid_diagnostic_end")) end
    -- CLI spans are inclusive. Neovim diagnostic ends are exclusive.
    local included = end_line:byte(end_start + 1)
    local width = included and (included >= 0xF0 and 4 or included >= 0xE0 and 3 or included >= 0xC0 and 2 or 1) or 0
    end_col = end_start + width
  end
  local bufnr = buffer_for(path)
  if not bufnr or dirty(bufnr) then return nil end
  local severity = ({ error = vim.diagnostic.severity.ERROR, warning = vim.diagnostic.severity.WARN,
    information = vim.diagnostic.severity.INFO, hint = vim.diagnostic.severity.HINT })[record.severity]
  return {
    bufnr = bufnr,
    diagnostic = {
      lnum = start_line_number - 1,
      col = start_col,
      end_lnum = end_line_number - 1,
      end_col = end_col,
      severity = severity,
      message = diagnostic_message(record),
      code = record.code,
      source = "recite-cli",
    },
  }
end

function M.new_namespace(name)
  return vim.api.nvim_create_namespace(name or "recite-cli")
end

function M.replace(namespace, records, project_root, known_buffers)
  if type(records) ~= "table" then error(protocol.error("invalid_diagnostics")) end
  local projected = {}
  for _, record in ipairs(records) do
    local value = to_diagnostic(record, project_root)
    if value then
      projected[value.bufnr] = projected[value.bufnr] or {}
      projected[value.bufnr][#projected[value.bufnr] + 1] = value.diagnostic
    end
  end
  vim.diagnostic.reset(namespace)
  for bufnr, diagnostics in pairs(projected) do
    vim.diagnostic.set(namespace, bufnr, diagnostics, { severity_sort = false })
    known_buffers[bufnr] = true
  end
  for bufnr in pairs(known_buffers) do
    if not projected[bufnr] then known_buffers[bufnr] = nil end
  end
end

function M.clear(namespace, known_buffers)
  vim.diagnostic.reset(namespace)
  for key in pairs(known_buffers) do known_buffers[key] = nil end
end

return M
